### PR TITLE
[android] Make expo-notifications installation ID provider in managed apps migrate legacy ID to a different location

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

During beta testing period I have been verifying if recent changes https://github.com/expo/expo/pull/11005 and https://github.com/expo/expo/pull/11019 work as we would expect them to. I noticed one of the scenarios, where we upgrade an SDK39 app to SDK40 and expect the `expo-notifications` installation ID to stay the same isn't true. I have investigated the issue (see [explanatory post](https://gist.github.com/sjchmiela/c9e8529dc4dd8ab425b8d0c33e7836d1)).

# How

In order to prepare for simple deprecation of `Constants.installationId`, `expo-notifications`' installation identifier migration now takes 3 different storage places into account in the following order:
1. `noBackupDir/expo_notifications_installation_uuid.txt` — the primary destination storage. If we have an ID there, this is the one we should use.
2. `SharedPreferences/UUID` — the legacy storage used by `expo-constants` and `expo-notifications` up to SDK40. If we have an ID there we are either:
    - in managed workflow where `SharedPreferences` are scoped, so this ID is most probably different from `Constants.installationId`
    - in bare workflow and `expo-constants` have not migrated the ID to `expo_installation_uuid.txt` file yet (maybe they're old?)
    
    
    If we detected an ID in this location we copy the value to our primary storage, but **do not remove the SharedPreferences entry**. While in the first scenario it would not cause any issues, in the second scenario if we removed the value `expo-constants`, when updated, would think it needs to create a new installation identifier, which we do not want.

    ⚠️ **Not removing `SharedPreferences` entry leaves app susceptible to the already existing bug** — if backup already has a `SharedPreferences` ID inside and we do not remove it, devices restored in the future will have the same ID. I see a couple of solutions to this problem:
      - keep using `expo_notifications_installation_uuid.txt ?? expo_installation_uuid.txt` for notifications' ID (as in original PR proposal)
      - remove `SharedPreferences` ID and guard against incompatible versions of dependencies with `peerDependencies` — a halfway solution, peer dependencies warnings are often ignored, in some projects may cause `Constants.installationId` to reset.
3. `noBackupDir/expo_installation_uuid.txt` — the legacy storage used by `expo-constants` and `expoview` since SDK40. If we have an ID there we are either (and we haven't already read ID from any of the aforementioned locations):
    - in managed workflow in an app that has not yet used `expo-notifications` (since there's no ID in scoped `SharedPreferences`)
    - in bare workflow where `expo-constants` have already migrated the ID.
    
    Either way we copy the value to our primary storage **not removing the file** not to cause the ID to change in other modules.

# Test Plan

I have confirmed that running an SDK40 app with these changes included, the ID gets migrated from scoped `SharedPreferences` to `expo_notifications_installation_id.txt`.

### CODE REVIEW

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java [hunk 1]
```diff
@@ -8,14 +8,16 @@
...
 import host.exp.exponent.di.NativeModuleDepsProvider;
+import host.exp.exponent.kernel.ExperienceId;
 import host.exp.exponent.storage.ExponentSharedPreferences;
 
 public class ScopedServerRegistrationModule extends ServerRegistrationModule {
   @Inject
   ExponentSharedPreferences mExponentSharedPreferences;
 
-  public ScopedServerRegistrationModule(Context context) {
+  public ScopedServerRegistrationModule(Context context, ExperienceId experienceId) {
```
Comment (cruzach) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java:18]: I think this means we need to update `detachedModuleRegistryAdapter` and pass in `experienceId` here- https://github.com/expo/expo/blob/%40sjchmiela/expo-notifications-managed-identifier/android/expoview/src/main/java/host/exp/exponent/experience/DetachedModuleRegistryAdapter.java#L63

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedInstallationId.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * A customization of {@link InstallationId} modifying location
+ * of the installation ID file to one that is *not* shared with
+ * other Expo modules. This makes expo-notifications installation
+ * ID provider maintain the same installation identifier
+ * in managed apps upgraded from SDK39 to SDK40. Without this change
+ * unscoped {@link InstallationId} would read the "shared"
+ * installation ID migrated by {@link ExponentInstallationId}
+ * or {@link host.exp.exponent.storage.ExponentSharedPreferences}
```
Comment (cruzach) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedInstallationId.java:21]: nice comment 👍

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java [hunk 3]
```diff
@@ -8,14 +8,16 @@
...
     super(context);
+    mInstallationId = new ScopedInstallationId(context);
```
Comment (cruzach) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java:20]: I believe this should be
```suggestion
    mInstallationId = new ScopedInstallationId(context, experienceId);
```

**Review by cruzach — CHANGES_REQUESTED** [review]

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java (hunk 3 above)
Comment (sjchmiela) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/notifications/ScopedServerRegistrationModule.java:19]: Of course! So sorry, this was the last change which was not supposed to break anything, etc.

I have removed it as we do not need it (see https://github.com/expo/expo/pull/11249/commits/5ae0e9315e4204966345a3104430aa5f64e3a8bb)! `ScopedContext` does it for us if we're *not* in a standalone app! 🎉

**Review by cruzach — APPROVED** [review]

**Review by ide — COMMENTED** [review]
I'd like to think about the future plans for these modules and IDs. Once `Constants.installationId` has been removed, we'll want remove its underlying classes. This means we probably do not want the notifications code to have logic like:
```
return read(expo_notifications_installation_id.txt) ?? read (expo_installation_uuid.txt)
```

This code would prevent us from deleting expo_installation_uuid.txt unless we migrated data away from it.

I think we should **always** write to expo_notifications_installation_id.txt, regardless of whether the ID came from shared or scoped preferences. This way we can just cleanly delete `expo_installation_uuid.txt` since the notifications code never reads/writes it.

**Review by ide — APPROVED** [review]

**Comment by ide** [comment]
I went ahead and merged this in as an exception since we want to unblock the Android release.

**Pull request merged**